synthesizer/src/block/transition/mod.rs

```diff
@@ -430,3 +430,31 @@ impl<N: Network> Transition<N> {
         self.finalize.into_iter().flatten()
     }
 }
+
+#[derive(Clone, Copy, PartialEq, Eq)]
+pub struct ProvingKeyId<N: Network> {
+    pub program_id: ProgramID<N>,
+    pub function_name: Identifier<N>,
+}
+
+impl<N: Network> Ord for ProvingKeyId<N> {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        // TODO: consider implementing Ord directly for Identifier and ProgramID.
+        let self_function_name_str = self.function_name.to_string();
+        let other_function_name_str = other.function_name.to_string();
+        let self_program_id_str = self.program_id.to_string();
+        let other_program_id_str = other.program_id.to_string();
+
+        // TODO: we should avoid the function_name from leaking into the program_id:
+        // option 1) '_' separator must be unique and not present in either string
+        // option 2) we can hash the function_name's and program_id's
+        (self_function_name_str + "_" + &self_program_id_str)
+            .cmp(&(other_function_name_str + "_" + &other_program_id_str))
```

Comment:
won't this be rather slow? I'd recommend relying on `Ord` of the members, or at least comparing bytes (as opposed to strings)

Reply:
Updated! 84a844611